client/auth.go

```diff
@@ -69,35 +73,51 @@ func (c *Conn) readInitialHandshake() error {
 	pos += 2
 
 	if len(data) > pos {
-		// skip server charset
+		// default server a_protocol_character_set, only the lower 8-bits
 		// c.charset = data[pos]
 		pos += 1
 
 		c.status = binary.LittleEndian.Uint16(data[pos : pos+2])
 		pos += 2
-		// capability flags (upper 2 bytes)
+
+		// The upper 2 bytes of the Capabilities Flags
 		c.capability = uint32(binary.LittleEndian.Uint16(data[pos:pos+2]))<<16 | c.capability
 		pos += 2
 
-		// auth_data is end with 0x00, min data length is 13 + 8 = 21
-		// ref to https://dev.mysql.com/doc/internals/en/connection-phase-packets.html#packet-Protocol::Handshake
-		maxAuthDataLen := 21
-		if c.capability&CLIENT_PLUGIN_AUTH != 0 && int(data[pos]) > maxAuthDataLen {
-			maxAuthDataLen = int(data[pos])
+		// length of the combined auth_plugin_data (scramble), if auth_plugin_data_len is > 0
+		authPluginDataLen := data[pos]
+		if (c.capability&CLIENT_PLUGIN_AUTH == 0) && (authPluginDataLen > 0) {
+			return errors.Errorf("invalid auth plugin data filler %d", authPluginDataLen)
 		}
+		pos++
 
 		// skip reserved (all [00])
-		pos += 10 + 1
+		pos += 6
 
-		// auth_data is end with 0x00, so we need to trim 0x00
-		resetOfAuthDataEndPos := pos + maxAuthDataLen - 8 - 1
-		c.salt = append(c.salt, data[pos:resetOfAuthDataEndPos]...)
+		// https://github.com/vapor/mysql-nio/blob/main/Sources/MySQLNIO/Protocol/MySQLProtocol%2BHandshakeV10.swift
+		if c.capability&CLIENT_LONG_PASSWORD != 0 {
+			// skip reserved (all [00])
+			pos += 4
+		} else {
+			// unknown
+			pos += 4
+		}
+
+		if rest := int(authPluginDataLen) - 8; rest > 0 {
```

Comment:
should we increase the rest to at least 13?

> Rest of the plugin provided data (scramble), $len=MAX(13, length of auth-plugin-data - 8)
https://dev.mysql.com/doc/dev/mysql-server/latest/page_protocol_connection_phase_packets_protocol_handshake_v10.html

Will take a look at free time later

Reply:
I do not have a reproduction of such a case, but yes, let it be

Reply:
Seems even if `rest := int(authPluginDataLen) - 8` and `rest < 0`, we should also increase `rest` to 13

Reply:
I looked into other implementations and changed the code a little. And I check compatibility on 5.7 and 8.0.